### PR TITLE
Add a flag `com.linecorp.armeria.annotatedServiceExceptionVerbosity`

### PR DESCRIPTION
**Pull request by hyangtack**
Motivation:
A user may want to know why his or her annotated service returns 400 or 500 status code from the log files.

Modifications:
- Add a flag `com.linecorp.armeria.annotatedServiceExceptionVerbosity`
- Add `ExceptionVerbosity` enum to control logging

Result:
A user can get every exception when adding `com.linecorp.armeria.annotatedServiceExceptionVerbosity=all` to the system properties.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 1]
```diff
@@ -523,6 +530,19 @@ public static int defaultMaxTotalAttempts() {
         return COMPOSITE_SERVICE_CACHE_SPEC;
     }
 
+    /**
+     * Returns the value of the {@code annotatedServiceExceptionLoggingMode} parameter which equals to
+     * one of the {@code all}, {@code unhandled} or {@code none}.
+     *
+     * <p>The default value of this flag is {@value DEFAULT_ANNOTATED_SERVICE_EXCEPTION_LOGGING_MODE}.
+     * Specify the
+     * {@code -Dcom.linecorp.armeria.annotatedServiceExceptionLoggingMode=<spec>} JVM option to override
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: `<spec>` -> `<mode>`?

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 2]
```diff
@@ -523,6 +530,19 @@ public static int defaultMaxTotalAttempts() {
         return COMPOSITE_SERVICE_CACHE_SPEC;
     }
 
+    /**
+     * Returns the value of the {@code annotatedServiceExceptionLoggingMode} parameter which equals to
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: Could you rephrase this to better explain what this flag does and why it is necessary?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java [hunk 3]
```diff
@@ -76,18 +77,21 @@
     private final ResponseConverterFunction providedResponseConverter;
 
     private final ResponseType responseType;
+    private final ExceptionLoggingMode exceptionLoggingMode;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java]: Perhaps no need to keep it as a field because it will never change?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java [hunk 4]
```diff
@@ -250,6 +254,12 @@ private HttpResponse convertResponse(ServiceRequestContext ctx, @Nullable Object
      */
     private HttpResponse convertException(RequestContext ctx, HttpRequest req, Throwable cause) {
         final Throwable peeledCause = Exceptions.peel(cause);
+
+        if (exceptionLoggingMode == ExceptionLoggingMode.ALL &&
+            logger.isWarnEnabled()) {
+            logger.warn("An exception from {}:", method.getName(), peeledCause);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java]: How about including the class name as well as the method name?

File: core/src/main/java/com/linecorp/armeria/server/annotation/DefaultExceptionHandler.java [hunk 5]
```diff
@@ -19,23 +19,51 @@
...
+        }
+
+        if (cause instanceof HttpResponseException) {
+            return ((HttpResponseException) cause).httpResponse();
+        }
+
+        if (exceptionLoggingMode == ExceptionLoggingMode.UNHANDLED &&
+            logger.isWarnEnabled()) {
+            logger.warn("No exception handler exists for the cause. " +
+                        DefaultExceptionHandler.class.getName() + " is handling it.",
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/DefaultExceptionHandler.java]: Could use `{}` instead of concatenation?

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 6]
```diff
@@ -523,6 +531,25 @@ public static int defaultMaxTotalAttempts() {
         return COMPOSITE_SERVICE_CACHE_SPEC;
     }
 
+    /**
+     * Returns the verbosity of exceptions logged by annotated HTTP services. If it is set to
+     * {@link ExceptionVerbosity#ALL}, all exceptions raised from annotated HTTP services are logged as
+     * {@code warn} level. If it is set to {@link ExceptionVerbosity#UNHANDLED}, exceptions, which are
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: `at {@code WARN} level`

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 7]
```diff
@@ -523,6 +531,25 @@ public static int defaultMaxTotalAttempts() {
         return COMPOSITE_SERVICE_CACHE_SPEC;
     }
 
+    /**
+     * Returns the verbosity of exceptions logged by annotated HTTP services. If it is set to
+     * {@link ExceptionVerbosity#ALL}, all exceptions raised from annotated HTTP services are logged as
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: Could be simplified using `<ul><li>`:

```
The value of this property is one of the following:
<ul>
  <li>{@link ExceptionVerbosity#ALL} - ...</li>
</ul>
```

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 8]
```diff
@@ -523,6 +531,25 @@ public static int defaultMaxTotalAttempts() {
...
+     * {@code warn} level. If it is set to {@link ExceptionVerbosity#UNHANDLED}, exceptions, which are
+     * not handled by {@link ExceptionHandler}s provided by a user and not well-known exceptions,
+     * would be logged as {@code warn} level. If it is set to {@link ExceptionVerbosity#NONE},
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: `at {@code WARN} level`

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 9]
```diff
@@ -190,6 +193,11 @@
             CSV_SPLITTER.splitToList(getNormalized(
                     "cachedHeaders", DEFAULT_CACHED_HEADERS, CharMatcher.ascii()::matchesAllOf));
 
+    private static final String DEFAULT_ANNOTATED_SERVICE_EXCEPTION_VERBOSITY = "unhandled";
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: `ExceptionVerbosity.UNHADLED.name()`?

Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/common/Flags.java]: It makes `{@value ...}` not work. So I intentionally used the string here. :-)
